sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/BulkExecutorTemp.java

```diff
@@ -0,0 +1,1138 @@
+            }
+        }
+        thresholds.recordSuccessfulOperation();
+        return Mono.just(ModelBridgeInternal.createCosmosBulkOperationResponse(
+            itemOperation,
+            cosmosBulkItemResponse,
+            actualContext));
+    }
+
+    private TContext getActualContext(CosmosItemOperation itemOperation) {
+        ItemBulkOperation<?, ?> itemBulkOperation = null;
+
+        if (itemOperation instanceof ItemBulkOperation<?, ?>) {
+            itemBulkOperation = (ItemBulkOperation<?, ?>) itemOperation;
+        }
+
+        if (itemBulkOperation == null) {
+            return this.batchContext;
+        }
+
+        TContext operationContext = itemBulkOperation.getContext();
+        if (operationContext != null) {
+            return operationContext;
+        }
+
+        return this.batchContext;
+    }
+
+    private Mono<CosmosBulkOperationResponse<TContext>> handleTransactionalBatchExecutionException(
+        CosmosItemOperation itemOperation,
+        Exception exception,
+        FluxSink<CosmosItemOperation> groupSink,
+        PartitionScopeThresholds thresholds) {
+
+        logger.debug(
+            "HandleTransactionalBatchExecutionException, PKRange {}, Error: {}, {}, Context: {} {}",
+            thresholds.getPartitionKeyRangeId(),
+            exception,
+            getItemOperationDiagnostics(itemOperation),
+            this.operationContextText,
+            getThreadInfo());
+
+        if (exception instanceof CosmosException && itemOperation instanceof ItemBulkOperation<?, ?>) {
+            CosmosException cosmosException = (CosmosException) exception;
+            ItemBulkOperation<?, ?> itemBulkOperation = (ItemBulkOperation<?, ?>) itemOperation;
+            if (preserveOrdering) {
+                CosmosItemIdentity cosmosItemIdentity = BulkExecutorUtil.createCosmosItemIdentity(itemOperation);
+                failedItems.compute(cosmosItemIdentity, (key, statusQueue) -> {
+                    if (statusQueue == null) {
+                        statusQueue = new StatusQueue<>();
+                    }
+                    statusQueue.add(itemOperation);
+                    return statusQueue;
+                });
+            }
+            // First check if it failed due to split, so the operations need to go in a different pk range group. So
+            // add it in the mainSink.
+
+            return itemBulkOperation.getRetryPolicy()
+                .shouldRetryForGone(cosmosException.getStatusCode(), cosmosException.getSubStatusCode(), itemBulkOperation, cosmosException)
+                .flatMap(shouldRetryGone -> {
+                    if (shouldRetryGone) {
+                        if (preserveOrdering) {
+                            CosmosItemIdentity cosmosItemIdentity = BulkExecutorUtil.createCosmosItemIdentity(itemOperation);
+                            failedItems.get(cosmosItemIdentity).setStatus(StatusQueue.Status.Split);
+                        }
```

Comment:
I am curious here - you set the status as `Split` for possibly other scenarios other than a partition split such as when the `PARTITION_KEY_RANGE_GONE` or the `COMPLETING_PARTITION_MIGRATION`. Is this to keep the behavior of adding the operation back to `mainSink` the expected behavior here?

Reply:
Yes its essentially for any error for when the cache is refreshed.